client/core/core.go

```diff
@@ -1629,6 +1629,80 @@ func (c *Core) WalletSettings(assetID uint32) (map[string]string, error) {
 	return dbWallet.Settings, nil
 }
 
+func (c *Core) setAppPass(pw []byte) (encrypt.Crypter, error) {
+	if len(pw) == 0 {
+		return nil, fmt.Errorf("empty password not allowed")
+	}
+
+	crypter := c.newCrypter(pw)
+	err := c.db.Store(keyParamsKey, crypter.Serialize())
+	if err != nil {
+		return nil, fmt.Errorf("error storing key parameters: %w", err)
+	}
+
+	return crypter, nil
+}
+
+func (c *Core) updateWalletsEncPW(oldCrypter, newCrypter encrypt.Crypter) error {
+	c.walletMtx.Lock()
+	defer c.walletMtx.Unlock()
+	for _, w := range c.wallets {
+		// If encrypted password set, decrypt using old app pw then encrypt again
+		// using new app pw.
+		if len(w.encPW) != 0 {
+			// Decrypt with old app pw.
+			pwB, err := oldCrypter.Decrypt(w.encPW)
+			if err != nil {
+				return err
+			}
+			// Re-encrypt with new app pw.
+			encPW, err := newCrypter.Encrypt(pwB)
+			if err != nil {
+				return err
+			}
+			// Store encrypted pw.
+			w.encPW = encPW
+			err = c.db.SetWalletPassword(w.dbID, w.encPW)
+			if err != nil {
+				return codedError(dbErr, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ChangeAppPass updates the application password to the provided new password
+// after validating the current password.
+func (c *Core) ChangeAppPass(appPW, newAppPW []byte) error {
+	// Validate current password.
+	oldCrypter, err := c.encryptionKey(appPW)
+	if err != nil {
+		return newError(authErr, "ChangePassApp password error: %v", err)
+	}
+	// Set new password.
+	newCrypter, err := c.setAppPass(newAppPW)
+	if err != nil {
+		return err
+	}
+	// Update xcWallets' encrypted passwords.
+	c.updateWalletsEncPW(oldCrypter, newCrypter)
+	// Update dexConnections' dexAccount private & encrypted key.
+	c.connMtx.Lock()
+	defer c.connMtx.Unlock()
+	for _, dc := range c.conns {
+		// Re-encrypt  private keys.
+		dc.acct.updateKeysEncryption(oldCrypter, newCrypter)
+		// Update account's db entry.
+		err = c.db.UpdateAccountKey(dc.acct.dbInfo())
```

Comment:
Initially I though `UpdateAccountKey` would be good if it were just targeting the key piece of the account data, but since it updates the entire account info, let's go with `UpdateAccount` with a doc on the interface that the only piece of account info that is likely to change is the encrypted representation of the private keys since changing the key pair itself effectively creates a new account.

Reply:
💯